### PR TITLE
Issue #377: assign a logic subset to logic (array)

### PR DESCRIPTION
**Pull request by RPG-coder-intc**
## Description & Motivation

Logic array subset assign

## Related Issue(s)

Issue #377 

## Testing

please see in logic_array_test.dart

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

### CODE REVIEW

File: lib/src/selection.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Copyright (C) 2023 Intel Corporation
```
Comment (mkorbel1) [inline lib/src/selection.dart:1]: Update copyrights to 2023-2024 :)

File: lib/src/signals/logic_array.dart [hunk 2]
```diff
@@ -127,4 +127,14 @@ class LogicArray extends LogicStructure {
     return LogicArray(dimensions, elementWidth,
         numUnpackedDimensions: numUnpackedDimensions, name: name);
   }
+
+  /// add doc later
+  void assignSubset(List<Logic> updatedSubset, int start) {
```
Comment (mkorbel1) [inline lib/src/signals/logic_array.dart:132]: did you mean to include this?  I think this was supposed to be in another PR for the other issue?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]
Looks good!  A couple small things then I think we can merge it in

File: lib/src/signals/logic_array.dart [hunk 3]
```diff
@@ -127,4 +127,34 @@ class LogicArray extends LogicStructure {
     return LogicArray(dimensions, elementWidth,
         numUnpackedDimensions: numUnpackedDimensions, name: name);
   }
+
+  /// Perform Assign operation on a Logic subset or slice
+  ///
+  /// Assigns part of this LogicArray with a given [updatedSubset] of type
+  /// [List<Logic>]. The update is performed from a given[start] position
```
Comment (mkorbel1) [inline lib/src/signals/logic_array.dart:134]: Missing a space after "given"

File: lib/src/signals/logic_array.dart [hunk 4]
```diff
@@ -127,4 +127,34 @@ class LogicArray extends LogicStructure {
...
+  /// sampleLogic.assignSubset(updatedSubset); // start = 0 by default
+  /// // assign updated subset to sampleLogic[10:10+updatedSubset.length]
+  /// sampleLogic.assignSubset(updatedSubset, 10);
+  /// ```
+  ///
+  void assignSubset(List<Logic> updatedSubset, {int start = 0}) {
+    // TO-DO later: optimize with dart: zip, reduce,
+
+    if (updatedSubset.length > elements.length - start) {
+      /* TODO: ERROR */
```
Comment (mkorbel1) [inline lib/src/signals/logic_array.dart:152]: TODO here should be fixed, optimization above is optional

File: test/logic_array_test.dart [hunk 5]
```diff
@@ -477,6 +477,18 @@ class IndexBitOfArrayModule extends Module {
   }
 }
 
+class AssignSubsetModule extends Module {
+  AssignSubsetModule(List<Logic> updatedSubset, {int? start}) {
+    final o = LogicArray([10], 3, name: 'logicArray');
+    o <= Const(LogicValue.ofString('1').replicate(10 * 3));
```
Comment (mkorbel1) [inline test/logic_array_test.dart:483]: we need to leave at least the signals not included in updatedSubset disconnected

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

File: lib/src/signals/logic_array.dart (hunk 4 above)
Comment (RPG-coder-intc) [reply lib/src/signals/logic_array.dart]: done

File: lib/src/signals/logic_array.dart [hunk 6]
```diff
@@ -127,4 +127,32 @@ class LogicArray extends LogicStructure {
...
+  /// List<Logic> updatedSubset;
+  /// // Assign part of sampleLogic as [updatedSubset]
+  /// sampleLogic.assignSubset(updatedSubset); // start = 0 by default
+  /// // assign updated subset to sampleLogic[10:10+updatedSubset.length]
+  /// sampleLogic.assignSubset(updatedSubset, 10);
+  /// ```
+  ///
+  void assignSubset(List<Logic> updatedSubset, {int start = 0}) {
+    if (updatedSubset.length > elements.length - start) {
+      throw LogicConstructionException('${updatedSubset.length}');
```
Comment (mkorbel1) [inline lib/src/signals/logic_array.dart:150]: Let's make this a `SignalWidthMismatchException` instead, and you may need to create a new constructor for it.

File: lib/src/exceptions/logic/signal_width_mismatch_exception.dart [hunk 7]
```diff
@@ -19,9 +19,8 @@ class SignalWidthMismatchException extends RohdException {
             ' $additionalMessage');
 
   /// Constructs a new [Exception] for when a dynamic has a wrong width.
-  SignalWidthMismatchException.forDynamic(
-      dynamic val, int expectedWidth, int actualWidth,
-      {String additionalMessage = ''})
+  SignalWidthMismatchException.forDynamic(int expectedWidth, int actualWidth,
+      {String additionalMessage = '', dynamic val = ''})
```
Comment (mkorbel1) [inline lib/src/exceptions/logic/signal_width_mismatch_exception.dart:23]: With this change, there will be 2 spaces between "Value" and "has" if `val` is `''`.

File: lib/src/signals/logic_array.dart [hunk 8]
```diff
@@ -127,4 +127,33 @@ class LogicArray extends LogicStructure {
...
+  /// List<Logic> updatedSubset;
+  /// // Assign part of sampleLogic as [updatedSubset]
+  /// sampleLogic.assignSubset(updatedSubset); // start = 0 by default
+  /// // assign updated subset to sampleLogic[10:10+updatedSubset.length]
+  /// sampleLogic.assignSubset(updatedSubset, 10);
+  /// ```
+  ///
+  void assignSubset(List<Logic> updatedSubset, {int start = 0}) {
+    if (updatedSubset.length > elements.length - start) {
+      throw SignalWidthMismatchException.forDynamic(
```
Comment (mkorbel1) [inline lib/src/signals/logic_array.dart:150]: It's not necessarily true that you expected the width to be exactly `elements.length - start`, it just needs to be less than that.  You could add a new constructor to the exception that allows you to write a custom message maybe?

File: test/logic_array_test.dart [hunk 9]
```diff
@@ -938,5 +953,45 @@ void main() {
       await SimCompare.checkFunctionalVector(mod, vectors);
       SimCompare.checkIverilogVector(mod, vectors);
     });
+
```
Comment (mkorbel1) [inline test/logic_array_test.dart:956]: We should add a test that it can hit the signal width mismatch exception

**Comment by RPG-coder-intc** [comment]
This seems to come from CI/CD
```
ERROR: 1 dead links found!
[✖] https://stackoverflow.com/ → Status: 403
```

**Comment by mkorbel1** [comment]
> @mkorbel1 This seems to come from CI/CD main?
> 
> ```
> ERROR: 1 dead links found!
> [✖] https://stackoverflow.com/ → Status: 403
> ```

Can you try updating this file to ignore it?  I wonder if StackOverflow is blocking GitHub for some reason
`.github/configs/mlc_config.json`

File: lib/src/exceptions/logic/signal_width_mismatch_exception.dart [hunk 10]
```diff
@@ -31,4 +31,12 @@ class SignalWidthMismatchException extends RohdException {
   SignalWidthMismatchException.forNull(dynamic val)
       : super('Could not infer width of input $val.'
             ' Please provide a valid width.');
+
+  /// Constructs a new [Exception] for when a dynamic has a wrong width.
+  SignalWidthMismatchException.forWidthOverflow(int actualWidth, int maxWidth,
+      {String customMessage = ''})
```
Comment (mkorbel1) [inline lib/src/exceptions/logic/signal_width_mismatch_exception.dart:37]: You could make it `String? customMessage` so that you can do `customMessage ?? 'Value has...'` instead of checking if it's an empty string

File: lib/src/exceptions/logic/signal_width_mismatch_exception.dart [hunk 11]
```diff
@@ -31,4 +31,12 @@ class SignalWidthMismatchException extends RohdException {
...
+      : super(customMessage == ''
+            ? 'Value has the wrong width.'
+                ' Expected $actualWidth to be less than $maxWidth.'
```
Comment (mkorbel1) [inline lib/src/exceptions/logic/signal_width_mismatch_exception.dart:40]: Less than or equal to?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

**Review by mkorbel1 — APPROVED** [review]
Looks great!

**Comment by mkorbel1** [comment]
> @mkorbel1 can this be merged?

Yes, I thought it was already, my bad!
